### PR TITLE
Export masks for  NSE-only constructs `.`, `J`, `patterns` and `measure`

### LINKED ISSUE #5277: patterns function missing
With both version 1.14.2 and dev version: Error: 'patterns' is not an exported object from 'namespace:data.table'

### LINKED ISSUE #5604: Export masks for all NSE-only constructs (currently at least '.', 'J', 'patterns')
This subsumes #5277.

We export masks for '.N`, `.I, `.GRP`, `.GRPI`, `.SD`, and `:=` to facilitate downstreams doing `@importFrom data.table .N` to hide `R CMD check` notes about undefined variables.

We should do the same for all the NSE constructs we interpret inside `[` / `melt()`. Currently, there's no export for `.`, `J`, and `patterns`, all of which could be masked as `function(...) NULL`.

### PR DESCRIPTION
**Pull request by Nj221102**
This pull request addresses issue #5604  and #5277 , which aims to export masks for all non-standard evaluation (NSE) only constructs within `[` and `melt()` functions. Currently, only `.N`, `.I`, `.GRP`, `.GRPI`,` .SD`, and `:=` are exported, but this PR extends the export functionality to include`.`,`J`, `measure` and `patterns`.

1. Masked NSE constructs as function(...) NULL.
2. exported the functions in `Namespace`.
3. changed one test .

This PR closes #5604
closes #5277

### CODE REVIEW

**Comment by tdhock** [comment]
please edit your first comment in this PR which says "This PR closes https://github.com/Rdatatable/data.table/issues/5604 https://github.com/Rdatatable/data.table/issues/5277." (two issues linked on one line)
please write one "Closes <issue>" per line so both issues will be closed when this PR is merged (right now only the first will be, because they are on the same line)
you can check that under Development in the right side of the screen (right now shows only one issue "Export..." but should show two)

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:21]: "NULL makes it clear ... that they're variables not functions" should probably be updated now, since some of these are functions.

**Review by tdhock — CHANGES_REQUESTED** [review]

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R:21]: should i just remove this line or we can replace it with something like "# The presence of `NULL` signals to R's mask check during loading that these are entities with specific behavior. However, it's worth noting that the inclusion of some new masks now encompasses functions as well."

Comment (Nj221102) [reply R/data.table.R]: I have updated the comment, if there is need of further changes for this PR plz suggest :)

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:18]: I don't think patterns and measure should be here, since they are defined in fmelt.R (which I guess is evaluated after data.table.R, over-writing this definition? and that is why the code still works / tests still pass?)

**Review by tdhock — CHANGES_REQUESTED** [review]

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R:18]: Yeah when i moved the masks of both measure and patterns to fmelt.R below the original functions,  a lot of test and 2 vignettes seems to cause issue, we might have to change a lot of tests and some code in vignette `datatable-reshape.Rmd, datatable-sd-usage.Rmd `, is that ok? and moving masks to fmelt.R is good enough or somewhere else will be better?

Comment (tdhock) [reply R/data.table.R:18]: no, just delete them, no mask needed since they are defined in fmelt.R

Comment (Nj221102) [reply R/data.table.R]: done 👍

**Comment by Nj221102** [comment]
@tdhock since their is no need of masks for `pattern` and `measure`, how should we go about  solving the problem in #5277 , should we export them regardless without masks or just change documentation to inform that these functions aren't being exported.

File: NAMESPACE
Comment (tdhock) [inline NAMESPACE:30]: please undo this change (patterns and measure should be exported)

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R]: these changes are good (patterns and measure should not be defined here)

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NAMESPACE
Comment (Nj221102) [reply NAMESPACE]: done 👍

**Comment by Nj221102** [comment]
done :)

File: NEWS.md
Comment (tdhock) [inline NEWS.md:99]: I find this confusing: Note: `measure` and `patterns` do not need additional masking as they are defined within `fmelt.R`, while the following constructs are masked as functions: `.`, `J`, `.N`, `.I`, `.GRP`, `.GRPI`, `.SD`, and `:=`

`.` and `J` and `:=` are functions but the others are not.
 
can it be deleted?  if not can you please explain why you need that / what value it adds?

Comment (Nj221102) [reply NEWS.md:99]: No lets delete that if its excessive and confusing, i only added these because i thought these are few highlights of the PR but yeah end user doesn't need to know this and will find it confusing.

deleting now.

File: NEWS.md
Comment (tdhock) [inline NEWS.md:99]: What is "Extended support" ?? can this first sentence be deleted? If not, why not? What value does it add?
"constructs like" implies to me that there may be others that are exported, that are not listed here, so can that be deleted too please? ( to avoid confusion) 
Also, news items typically start with the function/object name, can you please edit? How about `.`, `J`, `measure`, and `patterns` are exported...

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NEWS.md
Comment (Nj221102) [reply NEWS.md]: done 👍 , if needed further changes plz suggest.

File: NEWS.md
Comment (tdhock) [inline NEWS.md:99]: I don't understand "This update ensures that these constructs are consistently available, maintaining consistency and functionality across your data.table usage." 
Isn't this relevant for package developers? Is it true that after merging this PR, a package function which uses `.` will no longer have to define it inside the function, to avoid a CRAN NOTE? That could be mentioned.

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NEWS.md
Comment (Nj221102) [reply NEWS.md]: Yes i agree, that merging this PR will enable package developers to not worry about CRAN NOTE while using these construct and i have changed the news item accordingly, I am sorry about taking this much reviewing for a news item but i think these reviews have made sure the news item is relevant to the users, thanks.

File: NEWS.md
Comment (tdhock) [inline NEWS.md:99]: again, what is consistency? please either delete or clarify

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NEWS.md
Comment (Nj221102) [reply NEWS.md:99]: consistency here means that  the behaviour of `.`, `J`, `measure`, and `patterns` constructs after being exported is predictable and aligning with the previously exported constructs like .N, .I, .GRP, .GRPI, .SD, and :=, like before we had to define them inside their functions but didn't have to do that with others which were exported, so after this PR,  behaviour of all NSE constructs are consistent,  i think #5604 had the consistency tag for same reason.

Comment (tdhock) [reply NEWS.md:99]: great that makes sense, please edit the NEWS item to clarify that meaning of consistency.
how about "... are now exported for use withing [ and melt functions, for consistency with `.N` ... which were the only previous NSE exports."

Comment (Nj221102) [reply NEWS.md]: updated 👍

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:19]: why "during loading" ?? I find it confusing, please delete unless you have a strong reason to keep. 
why "aren't defined with specific behaviors" ?? each of these actually does have a specific behavior. please either delete or clarify.

**Review by tdhock — CHANGES_REQUESTED** [review]

File: NEWS.md
Comment (Nj221102) [reply NEWS.md]: done :)

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R:19]: Yeah lets delete that, to begin with this comment was their to notify that NULL signals that they're variables not functions, but now that we have function in same condition.

> why "during loading" ?

Same as the previous comment, `makes it clear (to the R's mask check on loading) `

>why "aren't defined with specific behaviors" ?

Yeah this seems to be a mistake by me, tbf i don't clearly remember what was going on when i wrote this but this seems wrong now.

Thanks for suggestions :)

File: inst/tests/tests.Rraw
Comment (tdhock) [inline inst/tests/tests.Rraw:2189]: please just remove the argument instead of using explicit NULL

**Review by tdhock — CHANGES_REQUESTED** [review]

File: inst/tests/tests.Rraw
Comment (Nj221102) [reply inst/tests/tests.Rraw:2189]: It gives an error 
```
Test 715 ran without errors but failed check that x equals y:
> x = J(a = 1:3, b = 4) 
First 0 of 0 (type 'NULL'): 
NULL
> y = TRUE 
First 1 of 1 (type 'logical'): 
[1] TRUE
target is NULL, current is logical
```
I was trying to find the issue and got to know that y arguement of test defaults to true, since our test ouputs Null it causes the test to fail,

```
test = function(num,x,y=TRUE,error=NULL,warning=NULL,message=NULL,output=NULL,notOutput=NULL,ignore.warning=NULL,options=NULL,env=NULL) {
```

Comment (tdhock) [reply inst/tests/tests.Rraw:2189]: I thought you were using NULL for the error arg, but actually it is the expected value (y arg).
so I guess that makes sense.
I wonder if it is desirable to keep that error though? What do you think @MichaelChirico @jangorecki  @ben-schwen ?

Comment (MichaelChirico) [reply inst/tests/tests.Rraw:2189]: Probably best to do something like with `:=` and instead of a stub returning `NULL`, make it a stub erroring "Hey you shouldn't call that here/there's probably a mistake":

https://github.com/Rdatatable/data.table/blob/e7b7e47280b34f1cb5323ffe9f78236a9307ff3c/R/data.table.R#L2768-L2771

Comment (Nj221102) [reply inst/tests/tests.Rraw:2189]: done :)

Comment (tdhock) [reply inst/tests/tests.Rraw:2189]: ok that is better but the error could be improved. it should say in what context J should be used.
also do we want a similar error for `.` ?

Comment (Nj221102) [reply inst/tests/tests.Rraw:2189]: Yes i think having a similar error for `.` won't do any harm, anyways these function aren't supposed to called outside of data.table operations anyways.

Comment (Nj221102) [reply inst/tests/tests.Rraw:2189]: > ok that is better but the error could be improved. it should say in what context J should be used. also do we want a similar error for `.` ?

done :)

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:16]: 'j' expression -> 'i' argument

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:20]: '.' is intended for use within data.table operations -> '.' can be used as an alias for 'list' within the square brackets of a data.table, DT[.]

**Review by tdhock — CHANGES_REQUESTED** [review]

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R]: done :)

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R]: done 👍

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:16]: argument

File: R/data.table.R
Comment (tdhock) [inline R/data.table.R:20]: please use consistent errors. in my opinion "the function should not be called here" is better (more specific) than "the symbol should not be used here"

**Review by tdhock — CHANGES_REQUESTED** [review]

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R]: done 👍

**Review by tdhock — APPROVED** [review]
ok looks good to me but please wait for a second review from @MichaelChirico @jangorecki or @ben-schwen

File: NAMESPACE
Comment (MichaelChirico) [inline NAMESPACE:30]: warning: I imagine this (`.`) will break some revdeps. Let's see.

File: NEWS.md
Comment (MichaelChirico) [inline NEWS.md:103]:
```suggestion
17. `.`, `J`, `measure`, and `patterns` are now exported for use within `[` and `melt()` functions, for consistency with`.N`, `.I`, `.GRP`, `.GRPI`, `.SD`, and `:=` which were the only previous NSE exports. This change means that package developers can simply import these names to avoid `R CMD check` `NOTE`s about these being undefined variables. See [#5604](https://github.com/Rdatatable/data.table/issues/5604) and [#5277](https://github.com/Rdatatable/data.table/issues/5277). Thanks to @tdhock for the suggestions and @Nj221102 for implementing.
```

File: R/data.table.R
Comment (MichaelChirico) [inline R/data.table.R:17]: restore this, it still applies to `.SD = ... = .EACHI`.

I would group these functions next to `:=` instead:

https://github.com/Rdatatable/data.table/blob/9be687206c2d640a889d9dc2d6592b613645ca45/R/data.table.R#L2768-L2771

Comment (Nj221102) [reply R/data.table.R]: done :)

File: NEWS.md
Comment (MichaelChirico) [inline NEWS.md:103]:
```suggestion
17. `.`, `J`, `measure`, and `patterns` are now exported for use within `[` and `melt()`, for consistency with other NSE exports like `.N` and `:=`, [#5604](https://github.com/Rdatatable/data.table/issues/5604). Package developers can now import these names to avoid `R CMD check` `NOTE`s about these being undefined variables. Thanks to @MichaelChirico and @ylelkes for the suggestions and @Nj221102 for implementing.
```

File: NEWS.md
Comment (MichaelChirico) [inline NEWS.md:103]:
```suggestion
17. `.`, `J`, `measure`, and `patterns` are now exported for use within `[` and `melt()`, for consistency with other NSE exports like `.N` and `:=`, [#5604](https://github.com/Rdatatable/data.table/issues/5604). Package developers can now import these names to avoid `R CMD check` `NOTE`s about them being undefined variables. Thanks to @MichaelChirico and @ylelkes for the suggestions and @Nj221102 for implementing.
```

File: R/data.table.R
Comment (MichaelChirico) [reply R/data.table.R]: Not sure how this comment got marked 'Outdated', but it still applies (ditto for `.()`)

Comment (Nj221102) [reply R/data.table.R]: I m not sure either, but i have updated it 👍

File: R/data.table.R
Comment (Nj221102) [reply R/data.table.R]: done 👍

**Review by MichaelChirico — APPROVED** [review]
Thank you!

**Pull request merged**
